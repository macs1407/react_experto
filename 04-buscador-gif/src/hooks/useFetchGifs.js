import { useState, useEffect } from "react"
import { getGifs } from "../helpers/getGifs";


// Un custom hook no es mas que una funcion
export const useFetchGifs = (categoria)=>{
    const [state, setState] = useState({
        data:[],
        loading:true,
    });

    // Se ejecuta cuando haya un cambio sobre una lista de dependencias si no se le envia nada en [] solo se ejecutara
    // Una vez cuando el componente se renderiza
    useEffect(() => {
        getGifs(categoria).then(imgs=>{
            setTimeout(()=>{
                setState({
                    data:imgs,
                    loading:false
                });
            },3000);
          
        });
    }, [categoria]);

    return state;
}