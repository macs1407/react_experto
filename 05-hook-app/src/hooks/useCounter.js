import { useState } from "react";

// Para crear un custom hook, hacemos lo siguiente
export const useCounter = (initialState = 10) => {
    // 1-) Establecer el useState
    const [counter, setCounter] = useState(initialState);

    const increment = ()=>{
        setCounter(counter + 1);
    }

    const decrement = ()=>{
        setCounter(counter - 1);
    }

    const reset = ()=>{
        setCounter(initialState);
    }
    // Lo que retorna el custom hook
    return {
        counter,
        increment,
        decrement,
        reset
    }
}
