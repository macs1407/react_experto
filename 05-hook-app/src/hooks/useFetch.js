import { useState, useEffect, useRef } from "react"

export const useFetch = (url) => {

    // Mantener la referencia a este hook cuando siga vivo o cuando el componente que lo usa esta montado
    const isMounted = useRef(true);

    const [state, setState] = useState({data: null, loading: true, error: null});

    // Solo se va a ejecutar una vez
    useEffect(() => {
        // y cuando se desmonte el useRef pasa a ser falso indicando que el componente ya no
        // esta montado
        return ()=>{
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {
        setState({loading:true,error:null,data:null});

        fetch(url).then(resp=>resp.json())
                  .then(data=>{
                      setTimeout(()=>{
                        // Si el comoponente aun esta montado hace el setState, esto haca por problemas de fuga
                        // de memoria y errores al tratar de montar el componente
                        if(isMounted.current){
                            setState({
                                loading:false,
                                error:false,
                                data
                            });
                        } else {
                            console.log('setState no se llamo');
                        }
                      },3000);                    
                   });
    }, [url]);
    return state;
}
