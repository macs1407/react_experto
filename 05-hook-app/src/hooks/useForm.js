import { useState } from "react"

export const useForm = (initialState={}) => {
    const [values, setValues] = useState(initialState);

    const reset = ()=>{
        setValues(initialState);
    }

    const handleInputChange = (e)=>{
         // Se haca una copia del estado actual con ... o si no sobre escrbiriamos el state actual
        // el [target.name] hace referencia a name y email
        setValues({
            ...values,
            [e.target.name]: e.target.value
        });
    }

    return [values, handleInputChange, reset];
}
