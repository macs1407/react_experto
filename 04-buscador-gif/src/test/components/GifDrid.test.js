import React from 'react';
import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import GifDrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas del componente GifDrid', () => {
    const categoria = "goku";
  
    test('Se debe renderizar el componente GifDrid', () => {
        useFetchGifs.mockReturnValue({
            data:[],
            loading:true
        });
        const wrapper = shallow(<GifDrid categoria={categoria}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar items cuando se cargan imÃ¡genes useFetchGifs', () => {
        
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        },
        {
            id: '123',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifDrid categoria={categoria}/>);
        
        //expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GridItem').length ).toBe( gifs.length );
    })
    
})
