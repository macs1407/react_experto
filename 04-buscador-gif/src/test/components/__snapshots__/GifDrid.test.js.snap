// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente GifDrid Se debe renderizar el componente GifDrid 1`] = `
<Fragment>
  <h2>
    goku
  </h2>
  <p>
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas del componente GifDrid debe de mostrar items cuando se cargan im√°genes useFetchGifs 1`] = `
<Fragment>
  <h2>
    goku
  </h2>
  <div
    className="card-grid"
  >
    <GridItem
      id="ABC"
      img={
        Object {
          "id": "ABC",
          "title": "Cualquier cosa",
          "url": "https://localhost/cualquier/cosa.jpg",
        }
      }
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
    <GridItem
      id="123"
      img={
        Object {
          "id": "123",
          "title": "Cualquier cosa",
          "url": "https://localhost/cualquier/cosa.jpg",
        }
      }
      key="123"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;
